File: src/main/java/com/dh/ondot/schedule/app/AlarmFacade.java
package com.dh.ondot.schedule.app;

import com.dh.ondot.member.domain.Member;
import com.dh.ondot.member.domain.service.MemberService;
import com.dh.ondot.schedule.domain.Schedule;
import com.dh.ondot.schedule.domain.repository.ScheduleRepository;
import com.dh.ondot.schedule.domain.service.RouteService;
import com.dh.ondot.schedule.domain.service.ScheduleService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class AlarmFacade {
    private final MemberService memberService;
    private final RouteService routeService;
    private final ScheduleService scheduleService;

    @Transactional(readOnly = true)
    public Schedule generateAlarmSettingByRoute(
            Long memberId, LocalDateTime appointmentAt,
            Double startX, Double startY, Double endX, Double endY
    ) {
        Member member = memberService.findExistingMember(memberId);

        Integer estimatedTime = routeService.calculateRouteTime(
                startX, startY,
                endX, endY
        );

        return scheduleService.createScheduleBasedOnMemberInfo(
                member, appointmentAt, estimatedTime
        );
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/application/AlarmFacade.kt
package com.dh.ondot.schedule.application

import com.dh.ondot.schedule.application.command.GenerateAlarmCommand
import com.dh.ondot.member.domain.service.MemberService
import com.dh.ondot.schedule.domain.Schedule
import com.dh.ondot.schedule.domain.service.AlarmService
import com.dh.ondot.schedule.domain.service.RouteService
import com.dh.ondot.schedule.domain.service.ScheduleService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AlarmFacade(
    private val memberService: MemberService,
    private val routeService: RouteService,
    private val scheduleService: ScheduleService,
    private val alarmService: AlarmService,
) {
    @Transactional
    fun generateAlarmSettingByRoute(memberId: Long, command: GenerateAlarmCommand): Schedule {
        val member = memberService.getMemberIfExists(memberId)

        val estimatedTimeMin = routeService.calculateRouteTime(
            command.startLongitude, command.startLatitude,
            command.endLongitude, command.endLatitude,
            command.transportType,
        )

        return scheduleService.setupSchedule(
            member, command.appointmentAt, estimatedTimeMin,
        )
    }

    fun recordAlarmTrigger(
        memberId: Long,
        alarmId: Long,
        scheduleId: Long,
        action: String,
        mobileType: String,
    ) {
        alarmService.recordTrigger(memberId, alarmId, scheduleId, action, mobileType)
    }
}
